Re-add ref and deref to RefCountable

adetaylor opened the pull request:
#### a3a7869bd3eb87e940114f1483902b54fdd8deba
<pre>
Re-add ref and deref to RefCountable
<a href="https://bugs.webkit.org/show_bug.cgi?id=303377">https://bugs.webkit.org/show_bug.cgi?id=303377</a>
<a href="https://rdar.apple.com/165212021">rdar://165212021</a>

Reviewed by Anne van Kesteren.

When actually used from Swift code (and only when actually used) this type
doesn&apos;t work, because SWIFT_SHARED_REFERENCE refers to functions in base
classes rather than the type itself. Add them directly to this type.

Canonical link: <a href="https://commits.webkit.org/303748@main">https://commits.webkit.org/303748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13da8d6f44ce77f8d5f7e24828817bef91d88d80

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133517 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6019 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44665 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141073 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 ios-apple 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6533 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5883 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102136 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85568 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | ⏳ 🛠 mac-apple 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136464 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4621 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/119694 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82937 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/4500 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2108 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125589 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113626 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37806 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143720 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132028 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5688 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38390 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/110516 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5770 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/161/builds/4876 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/110699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28054 "Built successfully and passed tests") | [  ~~🧪 vision-wk2~~](https://ews-build.webkit.org/#/builders/164/builds/4364 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115952 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5743 "Built successfully") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/34264 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164993 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5590 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69195 "Built successfully") | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43115 "Found 6 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/regexp-zero-length-alternatives.js.layout, stress/regexp-bm-search-character-non-fixed-size.js.dfg-eager, stress/regexp-named-capture-groups.js.dfg-eager, stress/to-lower-case-intrinsic-on-empty-rope.js.bytecode-cache, stress/to-lower-case-intrinsic-on-empty-rope.js.no-llint, stress/v8-regexp-strict.js.bytecode-cache (failure)") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/5832 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [❌ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5699 "Failed to checkout and rebase branch from PR 54687") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/RefCountable.h:
```diff
@@ -49,6 +49,18 @@ class RefCountable : public ThreadSafeRefCounted<RefCountable<T>> {
     }
 
+#ifdef __swift__
+    void ref()
```
Can this be `const` like our ref() / deref() functions usually are?

adetaylor replied:
Done

cdumez commented on Source/WTF/wtf/RefCountable.h:
```diff
@@ -49,6 +49,18 @@ class RefCountable : public ThreadSafeRefCounted<RefCountable<T>> {
+        ThreadSafeRefCounted<RefCountable<T>>::ref();
+    }
+
+    void deref()
```
Can this be `const` like our ref() / deref() functions usually are?

cdumez commented on Source/WTF/wtf/RefCountable.h:
```diff
@@ -49,6 +49,18 @@ class RefCountable : public ThreadSafeRefCounted<RefCountable<T>> {
+#ifdef __swift__
+    void ref()
+    {
+        ThreadSafeRefCounted<RefCountable<T>>::ref();
```
Wouldn't `ThreadSafeRefCounted::ref()` work?

adetaylor replied:
No, apparently because the superclass is templated.
> `error: unknown type name 'ThreadSafeRefCounted'; did you mean 'ThreadSafeRefCountedBase'?`

I was a bit surprised too.

cdumez commented on Source/WTF/wtf/RefCountable.h:
```diff
@@ -49,6 +49,18 @@ class RefCountable : public ThreadSafeRefCounted<RefCountable<T>> {
+
+    void deref()
+    {
+        ThreadSafeRefCounted<RefCountable<T>>::deref();
```
Wouldn't `ThreadSafeRefCounted::deref()` work?

adetaylor replied:
See comment above

cdumez commented on Source/WTF/wtf/RefCountable.h:
```diff
@@ -49,6 +49,18 @@ class RefCountable : public ThreadSafeRefCounted<RefCountable<T>> {
     {
     }
 
+#ifdef __swift__
```
This needs a FIXME comment and a radar because this looks like a huge hack.

adetaylor replied:
Done

adetaylor committed:
Re-add ref and deref to RefCountable
https://bugs.webkit.org/show_bug.cgi?id=303377
rdar://165212021

Reviewed by Anne van Kesteren.

When actually used from Swift code (and only when actually used) this type
doesn't work, because SWIFT_SHARED_REFERENCE refers to functions in base
classes rather than the type itself. Add them directly to this type.

Canonical link: https://commits.webkit.org/303748@main
